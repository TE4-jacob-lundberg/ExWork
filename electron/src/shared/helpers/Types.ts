export interface ILink {
  url: string;
  label: string;
}

export interface IGameFormFields {
  title: string;
  abbreviation: string;
  image: string; 
  xAxis: string;
  yAxis: string;
  links: ILink[];
  fileNames: string[];
}

export interface IGameFormErrors { [id: string]: string; }

export interface IGame {
  id: string;
  title: string;
  abbreviation: string | null;
  image?: IGameImage;
  links: ILink[];
  fileNames: string[];
}

export interface IGameImage {
  url: string;
  bannerPos: IPosition;
}

export interface IPosition {
  x: string;
  y: string;
}
