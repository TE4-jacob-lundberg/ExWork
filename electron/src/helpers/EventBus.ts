/* eslint-disable @typescript-eslint/no-explicit-any */
const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

export { EventBus };
