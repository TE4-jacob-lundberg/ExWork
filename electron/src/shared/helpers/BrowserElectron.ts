// eslint-disable-next-line @typescript-eslint/no-explicit-any
export let electron: any;

if (process.env.ENVIRONMENT === 'test') electron = require('electron');
else electron = window.require('electron');
