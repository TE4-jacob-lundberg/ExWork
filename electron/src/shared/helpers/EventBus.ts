/* eslint-disable @typescript-eslint/no-explicit-any */
interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

export { EventBus };
