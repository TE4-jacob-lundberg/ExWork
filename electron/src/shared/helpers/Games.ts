import { standardGames, IGame } from '../constants/standardGames';

export class Games {

  public static all(): IGame[] {
    if (!localStorage.getItem('games')) localStorage.setItem('games', JSON.stringify(standardGames));

    return JSON.parse(localStorage.getItem('games')!) || [];
  }

  public static addNew(game: IGame): string {
    const savedGames: IGame[] = JSON.parse(localStorage.getItem('games')!);
    if (savedGames.some(sGame => sGame.id === game.id)) return 'DUPLICATE_ID';
    
    localStorage.setItem('games', JSON.stringify(savedGames.concat(game)));
    return '';
  }
}
