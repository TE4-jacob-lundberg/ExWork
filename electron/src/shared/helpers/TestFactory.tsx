/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createBrowserHistory } from 'history';

export function factory(component: JSX.Element): RenderResult {
  return render(component);
}

export function routerFactory(component: JSX.Element, options: any = { history: createBrowserHistory() }): RenderResult {
  return factory(
    <Router history={options.history}>
      {component}
    </Router>
  );
}
