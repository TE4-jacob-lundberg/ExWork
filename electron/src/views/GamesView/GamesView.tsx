import React, { useState, useEffect, useCallback } from 'react';
import { css, jsx } from '@emotion/core'; // eslint-disable-line @typescript-eslint/no-unused-vars
import Styled from '@emotion/styled';
import { IpcRendererEvent } from 'electron';
import { Window } from 'node-window-manager';
import { useHistory } from 'react-router-dom';

import { electron } from '../../shared/helpers/BrowserElectron';
import { routes } from '../../shared/constants/routes';
import { IGame } from '../../shared/constants/standardGames';
import GameBannerComponent from './components/GameBannerComponent';
import ModalComponent from '../../shared/components/ModalComponent';
import ButtonComponent from '../../shared/components/ButtonComponent';
import { Games } from '../../shared/helpers/Games';

/* @jsx jsx */

const ipcRenderer = electron.ipcRenderer;


const ContainerStyled = Styled.div`
  height: 100vh;
  width: 100vw;
  position: relative;
  display: flex;
  align-items: center;
`;

type ContainerProps = {
  page: number;
  gameAmount: number;
}

const BannerContainerStyled = Styled.div<ContainerProps>`
  position: absolute;
  left: ${(props): number => -(86 * props.page)}vw;
  height: 70vh;
  margin: auto;
  display: grid;
  grid-template-columns: repeat(${(props): number => Math.ceil(props.gameAmount / 3)}, 1fr);
  grid-gap: 16vw;
  transition: 0.75s ease-in-out;
`;

const GameTrioContainerStyled = Styled.div`
  width: 70vw;
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-gap: 4px;
`;

const buttonStyling = css`
  position: absolute;
  z-index: 999;
`;

interface Props {}

const GamesView: React.FC<Props> = function () {
  const [runningGame, setRunningGame] = useState<IGame>({title: '', id: '', abbreviation: '', links: [], fileNames: []});
  const [showModal, setShowModal] = useState(false);
  const [page, setPage] = useState(0);
  const history = useHistory();
  const allGames = Games.all();

  useEffect(() => {
    if (!runningGame.id || localStorage.getItem('alreadyNotifiedOfRunningGame') === 'true') return;
    localStorage.setItem('alreadyNotifiedOfRunningGame', 'true');
    setShowModal(true);
  }, [runningGame]);

  const handleShowingWindows = useCallback((_event: IpcRendererEvent , prevWin: Window): void => {
    if (!prevWin.id) return;

    const prevAppFileName = prevWin.path.match(/(?<=\/)[^/]*\..*$/)![0];
    const identifiedGame = allGames.find(game => game.fileNames.includes(prevAppFileName));
    if (identifiedGame) setRunningGame(identifiedGame);
  }, [allGames]);

  useEffect(() => {
    ipcRenderer.on('showing-windows', handleShowingWindows);

    return (): void => {
      ipcRenderer.removeListener('showing-windows', handleShowingWindows);
    };
  }, [handleShowingWindows]);

  function GameBanners(): JSX.Element[] {
    const acc: JSX.Element[] = [];
    for (let index = 0; index < allGames.length; index += 3){ 
      const last = index + 3 <= allGames.length - 1 ? index + 3 : allGames.length;

      acc.push((
        <GameTrioContainerStyled key={index}>
          {allGames.slice(index, last).map(game => 
            <GameBannerComponent 
              game={game}
              key={game.title}
            />
          )}
        </GameTrioContainerStyled>));
    }
    return acc;
  }

  function handleNextPage(): void {
    setPage(page + 1);
  }

  function handlePrevPage(): void {
    setPage(page - 1);
  }

  return (
    <React.Fragment>
      {showModal && <ModalComponent 
        title="Recognized Game"
        body={`I recognized "${runningGame.title}" is running in the background. Do you want to see its links?`}
        onClose={(): void => setShowModal(false)}
        cancelAble
        onCancel={(): void => setShowModal(false)}
        onContinue={(): void => history.push(routes.gameLinks.replace(':gameID', runningGame.id))}
      />}
      <ButtonComponent
        onClick={handlePrevPage}
        iconSize="48px"
        disabled={page === 0}
        styling={css`
          left: -100px;
          ${buttonStyling.styles}
        `}
      >
        <i className="material-icons">
          chevron_left
        </i>
      </ButtonComponent>
      <ContainerStyled>
        <BannerContainerStyled page={page} gameAmount={allGames.length}>
          {GameBanners()}
        </BannerContainerStyled>
      </ContainerStyled>
      <ButtonComponent
        onClick={handleNextPage}
        iconSize="48px"
        disabled={page === Math.ceil(allGames.length / 3) - 1}
        styling={css`
          right: -100px;
          ${buttonStyling.styles}
        `}
      >
        <i className="material-icons">
          chevron_right
        </i>
      </ButtonComponent>
      <ButtonComponent
        onClick={(): void => history.push(routes.addGame)}
        iconSize="48px"
        title="Add your own game"
        styling={css`
          position: absolute;
          right: -100px;
          bottom: 0;
        `}
      >
        <i className="material-icons">
          add
        </i>
      </ButtonComponent>
    </React.Fragment>
  );
};

export default GamesView;
