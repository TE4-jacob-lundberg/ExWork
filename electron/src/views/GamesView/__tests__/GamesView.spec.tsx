import React from 'react';

import { routerFactory } from '../../../shared/helpers/TestFactory';
import GamesView from '../GamesView';

jest.mock('../../../shared/helpers/BrowserElectron.ts', () => {
  return {
    electron: {
      ipcRenderer: {
        on: jest.fn(),
        removeListener: jest.fn(),
      },
    },
  };
});

jest.mock('react-indexed-db', () => ({
  useIndexedDB: (): object => {
    return {
      getAll: (): Promise<[]> => Promise.resolve([]),
    };
  },
}));

describe('GamesView.tsx', () => {
  it('renders correctly', () => {
    const { container } = routerFactory(<GamesView />);

    expect(container).toMatchSnapshot();
  });
});
