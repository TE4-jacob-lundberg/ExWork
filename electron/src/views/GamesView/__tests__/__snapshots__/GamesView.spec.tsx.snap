// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GamesView.tsx renders correctly 1`] = `
<div>
  <button
    class="css-15orcq3"
    disabled=""
  >
    <i
      class="material-icons"
    >
      chevron_left
    </i>
  </button>
  <div
    class="css-1755h8c"
  >
    <div
      class="css-1vh095t"
    />
  </div>
  <button
    class="css-wg5tg3"
  >
    <i
      class="material-icons"
    >
      chevron_right
    </i>
  </button>
  <button
    class="css-1bs24ov"
    title="Add your own game"
  >
    <i
      class="material-icons"
    >
      add
    </i>
  </button>
</div>
`;
